> Concept ACK.

Thanks!

> This feels very deja-vu.
> Is it somehow possible to prevent these from being introduced? 

Heh, the deja-vu is understandable since you [asked](https://github.com/bitcoin/bitcoin/pull/9344#issuecomment-266971851) the same question ("Is it somehow possible to prevent these from being introduced?") in a gmaxwell PR with _exactly_ the same title four years ago: #9344 ("Do not run functions with necessary side-effects in assert()") :D

It can trivially be linted away if we want to. Let me know if we want such a linter: I'd be happy to write one if we have a consensus Concept ACK on that.

> How did you find them?

I stumbled upon these when reading static analyzer logs in search for more severe issues. Static analysers typically flag assertions with side effects: see #18920 for instructions on how to analyze Bitcoin Core using Clang Static Analysis, clang-tidy and cppcheck.

Example:

```
$ CPPCHECK_VERSION=2.3
$ curl -s https://codeload.github.com/danmar/cppcheck/tar.gz/${CPPCHECK_VERSION} | tar -zxf - --directory /tmp/
$ (cd /tmp/cppcheck-${CPPCHECK_VERSION}/ && make CFGDIR=/tmp/cppcheck-${CPPCHECK_VERSION}/cfg/ > /dev/null)
$ /tmp/cppcheck-${CPPCHECK_VERSION}/cppcheck \
    --enable=all \
    --language=c++ \
    --std=c++17 \
    -D__cplusplus \
    -DCLIENT_VERSION_BUILD \
    -DCLIENT_VERSION_IS_RELEASE \
    -DCLIENT_VERSION_MAJOR \
    -DCLIENT_VERSION_MINOR \
    -DCLIENT_VERSION_REVISION \
    -DCOPYRIGHT_YEAR \
    -DDEBUG \
    -DHAVE_WORKING_BOOST_SLEEP_FOR \
    -I src/ \
    -q \
    $(git ls-files -- "*.cpp" "*.h" ":(exclude)src/leveldb/" ":(exclude)src/secp256k1/" ":(exclude)src/univalue/") 2>&1 | \
    grep -A2 assertWithSideEffect

src/bench/chacha_poly_aead.cpp:34:21: warning: Assert statement calls a function which may have desired side effects: 'Crypt'. [assertWithSideEffect]
        assert(aead.Crypt(seqnr_payload, seqnr_aad, aad_pos, out.data(), out.size(), in.data(), buffersize, true));
                    ^
src/bench/chacha_poly_aead.cpp:38:25: warning: Assert statement calls a function which may have desired side effects: 'GetLength'. [assertWithSideEffect]
            assert(aead.GetLength(&len, seqnr_aad, aad_pos, in.data()));
                        ^
src/bench/chacha_poly_aead.cpp:39:25: warning: Assert statement calls a function which may have desired side effects: 'Crypt'. [assertWithSideEffect]
            assert(aead.Crypt(seqnr_payload, seqnr_aad, aad_pos, out.data(), out.size(), in.data(), buffersize, true));
                        ^
--
src/init.cpp:1666:40: warning: Assert statement calls a function which may have desired side effects: 'CanFlushToDisk'. [assertWithSideEffect]
                    assert(chainstate->CanFlushToDisk());
                                       ^
```

--- a/src/bench/chacha_poly_aead.cpp
+++ b/src/bench/chacha_poly_aead.cpp
@@ -31,12 +31,15 @@ static void CHACHA20_POLY1305_AEAD(benchmark::Bench& bench, size_t buffersize, b
     uint32_t len = 0;
     bench.batch(buffersize).unit("byte").run([&] {
         // encrypt or decrypt the buffer with a static key
-        assert(aead.Crypt(seqnr_payload, seqnr_aad, aad_pos, out.data(), out.size(), in.data(), buffersize, true));
+        const bool crypt_ok_1 = aead.Crypt(seqnr_payload, seqnr_aad, aad_pos, out.data(), out.size(), in.data(), buffersize, true);
+        assert(crypt_ok_1);
 
         if (include_decryption) {
             // if we decrypt, include the GetLength
-            assert(aead.GetLength(&len, seqnr_aad, aad_pos, in.data()));
-            assert(aead.Crypt(seqnr_payload, seqnr_aad, aad_pos, out.data(), out.size(), in.data(), buffersize, true));
+            const bool get_length_ok = aead.GetLength(&len, seqnr_aad, aad_pos, in.data());
+            assert(get_length_ok);
+            const bool crypt_ok_2 = aead.Crypt(seqnr_payload, seqnr_aad, aad_pos, out.data(), out.size(), in.data(), buffersize, true);
+            assert(crypt_ok_2);
         }
 
         // increase main sequence number

--- a/src/validation.h
+++ b/src/validation.h
@@ -564,7 +564,7 @@ public:
 
     //! @returns whether or not the CoinsViews object has been fully initialized and we can
     //!          safely flush this object to disk.
-    bool CanFlushToDisk() EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+    bool CanFlushToDisk() const EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
         return m_coins_views && m_coins_views->m_cacheview;
     }
 

--- a/src/wallet/test/coinselector_tests.cpp
+++ b/src/wallet/test/coinselector_tests.cpp
@@ -64,7 +64,8 @@ static void add_coin(CWallet& wallet, const CAmount& nValue, int nAge = 6*24, bo
     if (spendable) {
         CTxDestination dest;
         std::string error;
-        assert(wallet.GetNewDestination(OutputType::BECH32, "", dest, error));
+        const bool destination_ok = wallet.GetNewDestination(OutputType::BECH32, "", dest, error);
+        assert(destination_ok);
         tx.vout[nInput].scriptPubKey = GetScriptForDestination(dest);
     }
     if (fIsFromMe) {

